src/main/scala/freeeventsourcing/syntax/ProcessSyntax.scala

```diff
@@ -0,0 +1,218 @@
+      }
+    }
+
+    /** Option inside a firstOf. */
+    sealed trait SwitchPath {
+      type Event
+      type Result
+      val action: Await[BC, Event]
+      def effect(event: Event): ProcessMonad[Result]
+    }
+    object SwitchPath {
+      type Aux[E, R] = SwitchPath { type Event = E; type Result = R }
+    }
+    /** All options of a firstOf. */
+    sealed trait Switch[Paths <: HList] {
+      type Events <: Coproduct
+      type A <: Alternatives[BC]
+      type Result <: Coproduct
+      def effectFor(paths: Paths)(event: A#Events): ProcessMonad[Result]
+      def alternatives(paths: Paths): A
+    }
+    object Switch {
+      type Aux[P <: HList, R <: Coproduct] = Switch[P] { type Result = R }
+
+      implicit def end = new Switch[HNil] {
+        type A = FirstOf.Empty[BC]
+        type Events = CNil
+        type Result = CNil
+        def effectFor(paths: HNil)(event: CNil) = throw new AssertionError("Cannot be reached, the compiler is supposed to prevent that")
+        def alternatives(paths: HNil) = FirstOf.Empty()
+      }
+      implicit def head[E, R, T <: HList](implicit t: Switch[T]) = new Switch[SwitchPath.Aux[E, R] :: T] {
+        type A = Alternative[BC, E, t.A]
+        type P = SwitchPath.Aux[E, R]
+        type Events = P#Event :+: t.Events
+        type Result = P#Result :+: t.Result
+        def effectFor(paths: P :: T)(event: A#Events) = event match {
+          case Inl(myEvent)    ⇒ paths.head.effect(myEvent).map(Inl(_))
+          case Inr(otherEvent) ⇒ t.effectFor(paths.tail)(otherEvent).map(Inr(_))
+        }
+        def alternatives(paths: P :: T) = {
+          def alt[A, M <: Alternatives[BC]](await: Await[BC, A], more: M) = Alternative[BC, A, M](await, more)
+          alt(paths.head.action, t.alternatives(paths.tail))
+        }
+      }
+    }
+
+    final class FirstOfBuilder[A <: HList] private[ProcessSyntax] (paths: A) {
+      /** Await an event using a selector. The event must be from this bounded context. */
+      def on[S <: WithEventType: EventSelector: ValidSelector](selector: S) = new OnBuilder(selector)
+
+      /** Await event from a specific aggregate. */
+      def from[Id, A](aggregate: Id)(implicit afi: AggregateFromId[Id, BC#Aggregates]) = {
+        val aggregateType = afi.aggregate(boundedContext.aggregates)
+        new FromAggregateBuilder[afi.Out](aggregateType, aggregate)
+      }
+
+      /** Wait until the specified instant, then return unit. */
+      def timeout[R](when: Instant)(thenDo: ProcessMonad[R]): FirstOfBuilder[SwitchPath.Aux[Unit, R] :: A] = {
+        val path = new SwitchPath {
+          type Event = Unit
+          type Result = R
+          val action = WaitUntil[BC](when)
+          def effect(event: Unit) = thenDo
+
+        }
+        new FirstOfBuilder(path :: paths)
+      }
+
+      private[ProcessSyntax] def collect: A = paths
+
+      final class OnBuilder[S <: WithEventType: EventSelector: ValidSelector] private[FirstOfBuilder] (selector: S) {
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Class name does not match the regular expression '"[A-Z][A-Za-z]*"'](https://www.codacy.com/app/msiegenthaler/free-event-sourcing/file/2571398663/issues/source?bid=3298199&fileBranchId=3298202#l177)